### PR TITLE
Add Homebrew support

### PR DESCRIPTION
**Pull request by helpermethod**
- [x]  add JReleaser Templates for the Homebrew Tap Formula and README
- [x] enable Homebrew packager in JReleaser's configuration
- [x] create a Personal Access Token with repo permissions to use as `JRELEASER_HOMEBREW_GITHUB_TOKEN`
- [x] add `sdkman_selfupdate_feature` flag to disable `selfupdate` functionality (including `auto_update`)
- [x] keep `sdkman_selfupdate_enable` flag for backwards-compatibility
- [x] introduce `sdkman_auto_update` flag for enabling/disabling `auto_update` (replaces `sdkman_selfupdate_enable`)
- [x] add tests to make sure that any combination of old / new flags works as expected

### CODE REVIEW

**Comment by aalmiray** [comment]
Watch out for https://github.com/jreleaser/jreleaser/issues/702 when upgrading to JReleaser `v1.0.0-M2` or later. Templates will use `{{distributionExecutableUnix}}` instead of `{{distributionExecutable}}`.

**Comment by helpermethod** [comment]
@marc0der Removing the `sdkman-selfupdate.sh` file as a post-install step is probably a little bit extreme but simplifies certain things. I'm still not sure if this is the best way 😆.

Alternatively we could introduce another flag for disabling the `selfupdate` functionality, and hope that no one switches it on.

**Comment by helpermethod** [comment]
Hi @marc0der, I've added some tests to make sure that the old behaviour doesn't break.

File: src/main/bash/sdkman-init.sh [hunk 1]
```diff
@@ -156,6 +156,8 @@ if [[ -z "${sdkman_curl_retry_max_time}" ]]; then sdkman_curl_retry_max_time=60;
 # set curl to continue downloading automatically
 if [[ -z "${sdkman_curl_continue}" ]]; then sdkman_curl_continue=true; fi
 
+if [[ -z "$sdkman_selfupdate_feature" ]]; then sdkman_selfupdate_feature=true; fi
```
Comment (marc0der) [inline src/main/bash/sdkman-init.sh:159]: This should never happen if the config is added by the sdkman-hooks service in `install.sh` and `update.sh`. I think we can remove it.

File: src/test/groovy/sdkman/specs/SelfupdateSpec.groovy [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+
+		then:
+		verifyOutput(bash.output)
+
+		where:
+		selfupdateFeature | selfupdateEnabled | autoUpdateEnabled | verifyOutput
+		"true"            | "true"            | "true"            | { it.contains("ATTENTION: A new version of SDKMAN is available...") }
+		"true"            | "true"            | "false"           | { it.contains("ATTENTION: A new version of SDKMAN is available...") }
+		"true"            | "false"           | "true"            | { it.contains("ATTENTION: A new version of SDKMAN is available...") }
+		"false"           | "true"            | "true"            | { !it.contains("ATTENTION: A new version of SDKMAN is available...") }
```
Comment (marc0der) [inline src/test/groovy/sdkman/specs/SelfupdateSpec.groovy:91]: Awesome, I sometimes forget how great Spock is. Nice work!

**Review by marc0der — COMMENTED** [review]
Looks good so far. We shouldn't forget to add the new config to the sdkman-hooks service as mentioned.

File: src/main/bash/sdkman-init.sh (hunk 1 above)
Comment (helpermethod) [reply src/main/bash/sdkman-init.sh:159]: I thought about not exposing the flag to non-Homebrew users, but I can see that it makes sense to add it to the sdkman-hooks service.

Comment (marc0der) [reply src/main/bash/sdkman-init.sh]: Perhaps you can add a comment before to say why you've got this line here.

**Comment by helpermethod** [comment]
@marc0der With the hook updates promoted to stable, I wonder if

```
# Attempt upgrade after all is done
if [[ "$COMMAND" != "selfupdate" && "$sdkman_selfupdate_feature" == "true" && ("$sdkman_auto_update" == "true" || "$sdkman_selfupdate_enable" == "true") ]]; then
	__sdkman_auto_update "$SDKMAN_REMOTE_VERSION" "$SDKMAN_VERSION"
fi
```

the `sdkman_selfupdate_enable` check here is actually needed? I guess the selfupdate hook will take care of replacing it, right?

Would be nice if we could get rid of that OR expression.

**Comment by marc0der** [comment]
@helpermethod You are right. The selfupdate should take care of it, provided we deploy the hooks before the CLI. That said, I suggest leaving it in until both have been deployed, then removing it in a tiny follow-up PR once the dust has settled.

**Comment by helpermethod** [comment]
Too late, already removed it 😂

**Comment by marc0der** [comment]
@helpermethod You are too fast :laughing: You probably need to update the `SdkmanBashEnvBuilder` and `SelfupdateSpec` too which still refer to the old variable.

**Comment by helpermethod** [comment]
Removed the `selfupdate_enable` flag from the data table.

**Comment by helpermethod** [comment]
Oh, and before we finally do the release I need to drop the old SDKMAN! Homebrew repository.

**Comment by marc0der** [comment]
Okay, are you happy for me to merge that so we can test it in the beta channel?

**Comment by helpermethod** [comment]
Let's go!

File: src/test/groovy/sdkman/specs/SelfupdateSpec.groovy [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+		bash.execute("source $bootstrapScript")
+
+		when:
+		bash.execute("sdk version")
+
+		then:
+		verifyOutput(bash.output)
+
+		where:
+		selfupdateFeature | autoUpdateEnabled | verifyOutput
```
Comment (marc0der) [inline src/test/groovy/sdkman/specs/SelfupdateSpec.groovy:86]: Could we also test the negative cases here? It would be good to assert when `!it.contains("ATTENTION...")`.

Comment (helpermethod) [reply src/test/groovy/sdkman/specs/SelfupdateSpec.groovy:86]: I think there is at least one test for the negative case, no?

Comment (marc0der) [reply src/test/groovy/sdkman/specs/SelfupdateSpec.groovy:86]: Ah sorry, missed that. All good then!

**Comment by felipecrs** [comment]
This was a breaking change for me, sad it wasn't tagged as one. I think it should be avoided to perform breaking changes without bumping the major version (like from 5 to 6).

https://github.com/felipecrs/dotfiles/commit/5ba2f2b338282335b27969ebaf4c033aeabe4125

**Comment by helpermethod** [comment]
Sorry to hear 😥!

The `sdk selfupdate` command will automatically perform the migration, so for most users it should be a backwards-compatible change.

That's why we didn't bump the major version.

**Comment by felipecrs** [comment]
Got it. Well, https://github.com/sdkman/sdkman-cli/pull/1069 kind of prevents it (in the sense that if someone didn't touch the default configuration, it will be handled by the migration). But only in that case.

**Comment by felipecrs** [comment]
By the way, docs are missing the new `sdkman_auto_update`:

https://sdkman.io/usage#config
